# 范例：受控异常

<br>

受控异常的处理方式略有不同。

首先我要建立(或使用)一个合适的异常：

```java
class BalanceException extends Exception {}
```

然后，调整调用端如下：

```java
        try {
            account.withdraw(amount);
            doTheUsualThing();
        } catch (BalanceException e) {
            handleOverdrawn();
        }
```

接下来我要修改`withdraw()`函数，让它以异常表示错误状况：

```java
    void withdraw(int amount) throws BalanceException {
        if (amount > _balance) throw new BalanceException();
        _balance -= amount;
    }
```

这个过程的麻烦在于：我必须一次性修改所有调用者和被它们调用的函数，否则编译器会报错。

如果调用者很多，这个步骤就实在太大了，其中没有编译和测试的保障。

<br>

这种情况下，我可以借助一个临时中间函数。

我仍然从先前相同的情况出发：

```java
        if (account.withdraw(amount) == -1)
            handleOverdrawn();
        else doTheUsualThing();
class Account

    int withdraw(int amount) {
        if (amount > _balance)
            return -1;
        else {
            _balance -= amount;
            return 0;
        }
    }
```

首先，创建一个`newWithdraw()`函数，让它抛出异常：

```java
    void newWithdraw(int amount) throws BalanceException {
        if (amount > _balance) throw new BalanceException();
        _balance -= amount;
    }
```

完成以后，编译并测试。

现在我可以逐一将调用旧函数的地方改为调用新函数：

```java
        try {
            account.newWithdraw(amount);
            doTheUsualThing();
        } catch (BalanceException e) {
            handleOverdrawn();
        }
```

由于新旧两个函数都存在，所以每次修改后我都可以编译、测试。

所有调用者都修改完毕后，旧函数便可移除，并使用`Rename Method (273)`修改新函数名称，使它与旧函数相同。

<br>

